File: gamepad_osk-poc/renderer.py
"""Pygame renderer for the on-screen keyboard."""

import pygame


class Renderer:
    def __init__(self, surface, theme, unit_size, config):
        self.surface = surface
        self.theme = theme
        self.config = config
        self.unit = unit_size
        self.pad = config.padding

        # Auto-scale font: ~32% of unit_size, or explicit config value
        font_size = config.font_size if config.font_size > 0 else max(12, int(unit_size * 0.32))
        font_small_size = max(10, font_size - 4)

        pygame.font.init()
        font_names = ["DejaVu Sans", "Liberation Sans", "FreeSans", "Arial"]
        self.font = None
        self.font_small = None
        for name in font_names:
            path = pygame.font.match_font(name)
            if path:
                self.font = pygame.font.Font(path, font_size)
                self.font_small = pygame.font.Font(path, font_small_size)
                break
        if not self.font:
            self.font = pygame.font.SysFont(None, font_size + 4)
            self.font_small = pygame.font.SysFont(None, font_size)

    def draw(self, keyboard):
        """Draw the full keyboard state."""
        self.surface.fill(self.theme.bg)

        layout = keyboard.layout
        y_offset = self.pad

        for row_idx, row in enumerate(layout):
            x_offset = self.pad
            for col_idx, key in enumerate(row):
                is_cursor = (row_idx == keyboard.cursor_row and
                             col_idx == keyboard.cursor_col)
                self._draw_key(key, x_offset, y_offset, is_cursor, keyboard)
                x_offset += int(key.width * self.unit) + self.pad
            y_offset += self.unit + self.pad

        # Draw accent popup if active
        if keyboard.accent_popup is not None:
            self._draw_accent_popup(keyboard)

        # Draw modifier indicators at the top-right
        self._draw_modifier_status(keyboard)

    def _draw_key(self, key, x, y, is_cursor, keyboard):
        w = int(key.width * self.unit)
        h = self.unit
        rect = pygame.Rect(x, y, w, h)

        # Determine colors
        if is_cursor:
            bg = self.theme.highlight_bg
            border = self.theme.highlight_border
        elif key.is_modifier and self._is_modifier_active(key, keyboard):
            bg = self.theme.modifier_active_bg
            border = self.theme.highlight_border
        elif key.is_modifier:
            bg = self.theme.modifier_bg
            border = self.theme.key_border
        elif key.label.startswith("F") and key.label[1:].isdigit():
            bg = self.theme.fn_key_bg
            border = self.theme.key_border
        else:
            bg = self.theme.key_bg
            border = self.theme.key_border

        # Draw rounded rect
        pygame.draw.rect(self.surface, bg, rect, border_radius=6)
        pygame.draw.rect(self.surface, border, rect, width=1, border_radius=6)

        # Draw label
        label = keyboard.get_display_label(key)
        text_color = self.theme.key_text
        if is_cursor:
            # Brighter text on highlight
            text_color = (255, 255, 255)

        text_surf = self.font.render(label, True, text_color)
        text_rect = text_surf.get_rect(center=rect.center)
        self.surface.blit(text_surf, text_rect)

        # Draw shift label hint (small, top-right) for non-cursor keys with shift labels
        if key.shift_label and not key.is_modifier and not is_cursor:
            shift_visible = keyboard.shift_active ^ keyboard.caps_active
            if not shift_visible:
                hint = key.shift_label
                hint_surf = self.font_small.render(
                    hint, True, self.theme.modifier_text)
                hint_rect = hint_surf.get_rect(topright=(rect.right - 4, rect.top + 2))
                self.surface.blit(hint_surf, hint_rect)

    def _is_modifier_active(self, key, keyboard):
        if key.modifier_type == "shift":
            return keyboard.shift_active
        elif key.modifier_type == "caps":
            return keyboard.caps_active
        elif key.modifier_type == "ctrl":
            return keyboard.ctrl_active
        elif key.modifier_type == "alt":
            return keyboard.alt_active
        elif key.modifier_type == "meta":
            return keyboard.meta_active
        return False

    def _draw_accent_popup(self, keyboard):
        """Draw horizontal accent selector above the current key."""
        accents, selected = keyboard.accent_popup
        row = keyboard.layout[keyboard.cursor_row]

        # Find cursor key position
        x_offset = self.pad
        for i in range(keyboard.cursor_col):
            x_offset += int(row[i].width * self.unit) + self.pad

        key_y = self.pad + keyboard.cursor_row * (self.unit + self.pad)
        popup_y = key_y - self.unit - self.pad

        # Draw popup background
        total_w = len(accents) * (self.unit + self.pad) + self.pad
        popup_rect = pygame.Rect(x_offset, popup_y, total_w, self.unit + self.pad * 2)
        pygame.draw.rect(self.surface, self.theme.accent_popup_bg, popup_rect,
                         border_radius=8)
        pygame.draw.rect(self.surface, self.theme.highlight_border, popup_rect,
                         width=2, border_radius=8)

        # Draw each accent option
        ax = x_offset + self.pad
        for i, (label, codepoint) in enumerate(accents):
            arect = pygame.Rect(ax, popup_y + self.pad, self.unit, self.unit)
            if i == selected:
                pygame.draw.rect(self.surface, self.theme.accent_highlight_bg,
                                 arect, border_radius=6)
            text = self.font.render(label, True, self.theme.accent_popup_text)
            text_rect = text.get_rect(center=arect.center)
            self.surface.blit(text, text_rect)
            ax += self.unit + self.pad

    def _draw_modifier_status(self, keyboard):
        """Draw small modifier indicators."""
        indicators = []
        if keyboard.shift_active:
            indicators.append("SHIFT")
        if keyboard.caps_active:
            indicators.append("CAPS")
        if keyboard.ctrl_active:
            indicators.append("CTRL")
        if keyboard.alt_active:
            indicators.append("ALT")
        if keyboard.meta_active:
            indicators.append("SUPER")

        if not indicators:
            return

        x = self.surface.get_width() - self.pad
        y = self.pad
        for label in indicators:
            text = self.font_small.render(label, True, self.theme.modifier_active_bg)
            text_rect = text.get_rect(topright=(x, y))
            # Background pill
            pill = text_rect.inflate(10, 4)
            pygame.draw.rect(self.surface, self.theme.modifier_active_bg, pill,
                             border_radius=4)
            text = self.font_small.render(label, True, (255, 255, 255))
            text_rect = text.get_rect(center=pill.center)
            self.surface.blit(text, text_rect)
            x = pill.left - 6


def calculate_unit_size(layout, screen_width, config):
    """Auto-scale unit_size so keyboard is ~70% of screen width.

    If config.unit_size > 0, use it as-is (explicit override).
    If 0, compute from screen resolution.
    """
    if config.unit_size > 0:
        return config.unit_size

    # Find the widest row in key-width units
    pad = config.padding
    max_units = 0
    max_keys = 0
    for row in layout:
        row_units = sum(k.width for k in row)
        if row_units > max_units:
            max_units = row_units
            max_keys = len(row)

    # Target 70% of screen width
    target_width = int(screen_width * 0.70)
    # Solve for unit: target = pad + sum(key.width * unit + pad) per key
    # ≈ (max_keys + 1) * pad + max_units * unit
    num_gaps = max_keys + 1
    unit = max(30, (target_width - num_gaps * pad) / max_units)
    return int(unit)


def calculate_window_size(layout, unit_size, padding):
    """Calculate window dimensions from layout and resolved unit_size."""
    max_row_width = 0
    for row in layout:
        row_width = padding
        for key in row:
            row_width += int(key.width * unit_size) + padding
        max_row_width = max(max_row_width, row_width)

    height = padding + len(layout) * (unit_size + padding)
    return max_row_width, height

File: renderer.go
package main

import (
	"errors"
	"os"
	"strings"
	"time"
)

const promptFontPath = "/usr/share/fonts/TTF/promptfont.ttf"

type Renderer struct {
	renderer       *SDLRenderer
	theme          Theme
	unit           int32
	pad            int32
	statusH        int32 // height of modifier status bar
	font           *Font
	fontSmall      *Font
	fontGlyph      *Font
	flashText string
	flashEnd  time.Time
}

func NewRenderer(r *SDLRenderer, theme Theme, unitSize, padding int32) (*Renderer, error) {
	fontSize := max32(12, int32(float64(unitSize)*0.32))
	glyphSize := max32(12, int32(float64(unitSize)*0.28))

	if err := TTF3Init(); err != nil {
		return nil, err
	}

	fontPath := findFont("DejaVu Sans", "Liberation Sans", "FreeSans")
	if fontPath == "" {
		return nil, errors.New("no suitable font found")
	}

	font, err := TTF3OpenFont(fontPath, float32(fontSize))
	if err != nil {
		return nil, err
	}
	fontSmall, err := TTF3OpenFont(fontPath, float32(max32(10, fontSize-4)))
	if err != nil {
		return nil, err
	}

	var fontGlyph *Font
	if _, err := os.Stat(promptFontPath); err == nil {
		fontGlyph, _ = TTF3OpenFont(promptFontPath, float32(glyphSize))
	}
	if fontGlyph == nil {
		fontGlyph, _ = TTF3OpenFont(fontPath, float32(max32(10, fontSize-6)))
	}

	return &Renderer{
		renderer:  r,
		theme:     theme,
		unit:      unitSize,
		pad:       padding,
		statusH:   max32(20, int32(float64(unitSize)*0.4)),
		font:      font,
		fontSmall: fontSmall,
		fontGlyph: fontGlyph,
	}, nil
}

func (r *Renderer) Draw(kb *KeyboardState) {
	t := r.theme
	setColor(r.renderer, t.Bg)
	SDL3RenderClear(r.renderer)

	// Draw modifier status bar
	r.drawModifierBar(kb)

	// Draw keyboard rows
	y := r.pad + r.statusH
	for ri, row := range kb.Layout {
		x := r.pad
		for ci, key := range row {
			isCursor := ri == kb.CursorRow && ci == kb.CursorCol
			r.drawKey(key, x, y, isCursor, kb)
			x += int32(key.Width*float64(r.unit)) + r.pad
		}
		y += r.unit + r.pad
	}

	// Accent popup
	if kb.AccentPopup != nil {
		r.drawAccentPopup(kb)
	}

	SDL3RenderPresent(r.renderer)
}

func (r *Renderer) drawModifierBar(kb *KeyboardState) {
	// Modifier pills (left-aligned)
	var labels []string
	if kb.ShiftActive {
		labels = append(labels, "SHIFT")
	}
	if kb.CapsActive {
		labels = append(labels, "CAPS")
	}
	if kb.CtrlActive {
		labels = append(labels, "CTRL")
	}
	if kb.AltActive {
		labels = append(labels, "ALT")
	}
	if kb.MetaActive {
		labels = append(labels, "SUPER")
	}

	x := r.pad
	tc := r.theme.ModifierActiveText
	for _, label := range labels {
		r.drawPill(label, x, r.pad, r.theme.ModifierActiveBg, tc)
		tw, _ := TTF3GetStringSize(r.fontSmall, label)
		x += tw + 10 + 6
	}

	// Status flash (right-aligned, 2 seconds)
	if r.flashText != "" && time.Now().Before(r.flashEnd) {
		sw, _ := SDL3GetRenderOutputSize(r.renderer)
		surf, err := TTF3RenderTextBlended(r.fontSmall, r.flashText, r.theme.KeyText)
		if err == nil {
			tex := SDL3CreateTextureFromSurface(r.renderer, surf)
			if tex != nil {
				surfW := SDL3SurfaceWidth(surf)
				surfH := SDL3SurfaceHeight(surf)
				dst := FRect{X: float32(sw - surfW - r.pad), Y: float32(r.pad + 2), W: float32(surfW), H: float32(surfH)}
				SDL3RenderTexture(r.renderer, tex, nil, &dst)
				SDL3DestroyTexture(tex)
			}
			SDL3DestroySurface(surf)
		}
	}
}

func (r *Renderer) drawPill(label string, x, y int32, bg Color, fg Color) {
	surf, err := TTF3RenderTextBlended(r.fontSmall, label, fg)
	if err != nil {
		return
	}
	surfW := SDL3SurfaceWidth(surf)
	surfH := SDL3SurfaceHeight(surf)
	defer SDL3DestroySurface(surf)
	pw := surfW + 10
	ph := surfH + 4
	pill := FRect{X: float32(x), Y: float32(y), W: float32(pw), H: float32(ph)}
	setColor(r.renderer, bg)
	SDL3RenderFillRect(r.renderer, pill)
	tex := SDL3CreateTextureFromSurface(r.renderer, surf)
	if tex != nil {
		dst := FRect{X: float32(x + 5), Y: float32(y + 2), W: float32(surfW), H: float32(surfH)}
		SDL3RenderTexture(r.renderer, tex, nil, &dst)
		SDL3DestroyTexture(tex)
	}
}

func (r *Renderer) drawKey(key KeyDef, x, y int32, isCursor bool, kb *KeyboardState) {
	w := int32(key.Width * float64(r.unit))
	h := r.unit
	rect := FRect{X: float32(x), Y: float32(y), W: float32(w), H: float32(h)}
	t := r.theme

	flashed := kb.IsFlashed(key)
	var bg, border Color
	switch {
	case isCursor || flashed:
		bg, border = t.HighlightBg, t.HighlightBorder
	case key.IsModifier && isModActive(key, kb):
		bg, border = t.ModifierActiveBg, t.HighlightBorder
	case key.IsModifier:
		bg, border = t.ModifierBg, t.KeyBorder
	case strings.HasPrefix(key.Label, "F") && len(key.Label) > 1 && key.Label[1] >= '0' && key.Label[1] <= '9':
		bg, border = t.FnKeyBg, t.KeyBorder
	default:
		bg, border = t.KeyBg, t.KeyBorder
	}

	// Fill
	setColor(r.renderer, bg)
	SDL3RenderFillRect(r.renderer, rect)
	// Border
	setColor(r.renderer, border)
	SDL3RenderRect(r.renderer, rect)

	// Label
	label := kb.DisplayLabel(key)
	tc := t.KeyText
	if isCursor || flashed {
		tc = Color{R: 255, G: 255, B: 255, A: 255}
	} else if key.IsModifier && isModActive(key, kb) {
		tc = t.ModifierActiveText
	}
	r.renderText(r.font, label, tc, rect, AlignCenter)

	// Shift hint (top-right)
	if key.ShiftLabel != "" && !key.IsModifier && !isCursor {
		if kb.ShiftActive == kb.CapsActive {
			r.renderText(r.fontSmall, key.ShiftLabel, t.ModifierText,
				FRect{X: float32(x), Y: float32(y), W: float32(w - 4), H: float32(h)}, AlignTopRight)
		}
	}

	// Controller glyph (bottom-right)
	if glyph, ok := KeyGlyphs[key.Code]; ok && r.fontGlyph != nil {
		r.renderText(r.fontGlyph, glyph, t.GlyphColor,
			FRect{X: float32(x), Y: float32(y), W: float32(w - 3), H: float32(h - 2)}, AlignBottomRight)
	}
}

func (r *Renderer) drawAccentPopup(kb *KeyboardState) {
	accents := kb.AccentPopup.Accents
	sel := kb.AccentPopup.Selected
	row := kb.Layout[kb.CursorRow]

	xo := r.pad
	for i := range kb.CursorCol {
		xo += int32(row[i].Width*float64(r.unit)) + r.pad
	}
	ky := r.pad + r.statusH + int32(kb.CursorRow)*(r.unit+r.pad) //nolint:gosec // G115: cursor index fits in int32
	py := ky - r.unit - r.pad

	tw := int32(len(accents))*(r.unit+r.pad) + r.pad //nolint:gosec // G115: accent count fits in int32
	pr := FRect{X: float32(xo), Y: float32(py), W: float32(tw), H: float32(r.unit + r.pad*2)}
	setColor(r.renderer, r.theme.AccentPopupBg)
	SDL3RenderFillRect(r.renderer, pr)
	setColor(r.renderer, r.theme.HighlightBorder)
	SDL3RenderRect(r.renderer, pr)

	ax := xo + r.pad
	for i, accent := range accents {
		ar := FRect{X: float32(ax), Y: float32(py + r.pad), W: float32(r.unit), H: float32(r.unit)}
		if i == sel {
			setColor(r.renderer, r.theme.AccentHighlightBg)
			SDL3RenderFillRect(r.renderer, ar)
		}
		r.renderText(r.font, accent.Label, r.theme.AccentPopupText, ar, AlignCenter)
		ax += r.unit + r.pad
	}
}

type TextAlign int

const (
	AlignCenter TextAlign = iota
	AlignTopRight
	AlignBottomRight
)

func (r *Renderer) renderText(font *Font, text string, color Color, rect FRect, align TextAlign) {
	if text == "" || font == nil {
		return
	}
	surf, err := TTF3RenderTextBlended(font, text, color)
	if err != nil {
		return
	}
	defer SDL3DestroySurface(surf)

	tex := SDL3CreateTextureFromSurface(r.renderer, surf)
	if tex == nil {
		return
	}
	defer SDL3DestroyTexture(tex)

	surfW := float32(SDL3SurfaceWidth(surf))
	surfH := float32(SDL3SurfaceHeight(surf))

	var dst FRect
	switch align {
	case AlignCenter:
		dst = FRect{
			X: rect.X + (rect.W-surfW)/2,
			Y: rect.Y + (rect.H-surfH)/2,
			W: surfW, H: surfH,
		}
	case AlignTopRight:
		dst = FRect{
			X: rect.X + rect.W - surfW,
			Y: rect.Y + 2,
			W: surfW, H: surfH,
		}
	case AlignBottomRight:
		dst = FRect{
			X: rect.X + rect.W - surfW,
			Y: rect.Y + rect.H - surfH,
			W: surfW, H: surfH,
		}
	}
	SDL3RenderTexture(r.renderer, tex, nil, &dst)
}

func (r *Renderer) SetTheme(t Theme) {
	r.theme = t
	r.Flash(t.Name)
}

func (r *Renderer) Flash(text string) {
	r.flashText = text
	r.flashEnd = time.Now().Add(2 * time.Second)
}

func (r *Renderer) Destroy() {
	if r.font != nil {
		TTF3CloseFont(r.font)
	}
	if r.fontSmall != nil {
		TTF3CloseFont(r.fontSmall)
	}
	if r.fontGlyph != nil {
		TTF3CloseFont(r.fontGlyph)
	}
}

func CalcUnitSize(layout [][]KeyDef, screenWidth int32, cfg Config) int32 {
	if cfg.Keys.UnitSize > 0 {
		return int32(cfg.Keys.UnitSize) //nolint:gosec // G115: unit size fits in int32
	}
	var maxUnits float64
	var maxKeys int
	for _, row := range layout {
		ru := 0.0
		for _, k := range row {
			ru += k.Width
		}
		if ru > maxUnits {
			maxUnits = ru
			maxKeys = len(row)
		}
	}
	scale := float64(cfg.Keys.Scale) / 100.0
	if scale <= 0 {
		scale = 0.70
	}
	target := float64(screenWidth) * scale
	pad := float64(cfg.Keys.Padding)
	unit := (target - float64(maxKeys+1)*pad) / maxUnits
	if unit < 30 {
		unit = 30
	}
	return int32(unit)
}

func CalcWindowSize(layout [][]KeyDef, unit, pad, statusH int32) (int32, int32) {
	var maxW int32
	for _, row := range layout {
		rw := pad
		for _, key := range row {
			rw += int32(key.Width*float64(unit)) + pad
		}
		if rw > maxW {
			maxW = rw
		}
	}
	h := pad + statusH + int32(len(layout))*(unit+pad) //nolint:gosec // G115: layout row count fits in int32
	return maxW, h
}

func setColor(r *SDLRenderer, c Color) {
	SDL3SetRenderDrawColor(r, c)
}

func isModActive(key KeyDef, kb *KeyboardState) bool {
	switch key.ModifierType {
	case "shift":
		return kb.ShiftActive
	case "caps":
		return kb.CapsActive
	case "ctrl":
		return kb.CtrlActive
	case "alt":
		return kb.AltActive
	case "meta":
		return kb.MetaActive
	}
	return false
}

func findFont(names ...string) string {
	// Common font directories on Linux
	dirs := []string{
		"/usr/share/fonts/TTF",
		"/usr/share/fonts/truetype/dejavu",
		"/usr/share/fonts/truetype/liberation",
		"/usr/share/fonts/truetype/freefont",
		"/usr/share/fonts/truetype",
		"/usr/share/fonts",
	}
	filePatterns := map[string][]string{
		"DejaVu Sans":    {"DejaVuSans.ttf"},
		"Liberation Sans": {"LiberationSans-Regular.ttf"},
		"FreeSans":        {"FreeSans.ttf"},
	}

	for _, name := range names {
		patterns := filePatterns[name]
		for _, dir := range dirs {
			for _, pat := range patterns {
				path := dir + "/" + pat
				if _, err := os.Stat(path); err == nil {
					return path
				}
			}
		}
	}
	return ""
}

func max32(a, b int32) int32 {
	if a > b {
		return a
	}
	return b
}
